File: .claude/skills/tiktok-lookup/scripts/bot.mjs
/**
 * bot.mjs — TikTok Similar Creator iMessage Bot
 *
 * Listens for any iMessage containing a TikTok profile URL,
 * looks up similar creators via TikHub API, and replies with profile links.
 *
 * Run: node bot.mjs
 * Requires: @photon-ai/imessage-kit (already in send-imessage/node_modules)
 */

import { createRequire } from 'module';
import { execFileSync, execFile } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, join, resolve } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SKILL_DIR = resolve(__dirname, '..');
const LOOKUP_PY = join(__dirname, 'lookup.py');
const ENV_PATH = resolve(SKILL_DIR, '..', '..', '.env');
const DATA_DIR = resolve(SKILL_DIR, '..', '..', '..', 'data');
const SCOUT_PY = resolve(SKILL_DIR, '..', 'scout-api', 'scripts', 'scout.py');
const KEYWORDS_HELPER_PY = resolve(SKILL_DIR, '..', 'scout-api', 'scripts', 'keywords_helper.py');
const CAMPAIGNS_DIR = resolve(SKILL_DIR, '..', '..', '..', 'context', 'campaigns');

// Load .env manually (no dotenv dep needed)
import { readFileSync, existsSync } from 'fs';
if (existsSync(ENV_PATH)) {
  for (const line of readFileSync(ENV_PATH, 'utf8').split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
    const [k, ...rest] = trimmed.split('=');
    process.env[k.trim()] ??= rest.join('=').trim();
  }
}

const require = createRequire(import.meta.url);
const PHOTON_PATH = resolve(
  __dirname,
  '../../../../send-imessage/node_modules/@photon-ai/imessage-kit'
);
const { IMessageSDK } = require(PHOTON_PATH);

const TIKTOK_URL_RE = /tiktok\.com\/(?:@([\w.]+)|t\/([\w]+))/i;
const SCOUT_RE = /^scout\s+#(\w+)(?:\s+(.+))?$/i;

import { readdirSync, statSync } from 'fs';

/** Resolve campaign name (case-insensitive). Returns canonical name or null. */
function resolveCampaign(rawName) {
  try {
    const dirs = readdirSync(CAMPAIGNS_DIR).filter(d => {
      try { return statSync(`${CAMPAIGNS_DIR}/${d}`).isDirectory(); } catch { return false; }
    });
    return dirs.find(d => d.toLowerCase() === rawName.toLowerCase()) || null;
  } catch { return null; }
}

/** List all campaign folder names. */
function listCampaigns() {
  try {
    return readdirSync(CAMPAIGNS_DIR).filter(d => {
      try { return statSync(`${CAMPAIGNS_DIR}/${d}`).isDirectory(); } catch { return false; }
    });
  } catch { return []; }
}

/** Get pending keyword count for a campaign via keywords_helper.py */
function getPendingCount(campaign) {
  try {
    const out = execFileSync('uv', ['run', KEYWORDS_HELPER_PY, 'pending', campaign], {
      timeout: 10000, env: { ...process.env },
    }).toString().trim();
    return parseInt(out) || 0;
  } catch { return 0; }
}

/** Spawn scout.py detached (fire-and-forget). scout.py sends its own iMessage updates. */
function spawnScout(campaign, keyword) {
  const args = keyword
    ? ['run', SCOUT_PY, campaign, keyword]
    : ['run', SCOUT_PY, campaign];
  const child = execFile('uv', args, {
    env: { ...process.env },
    timeout: 0,  // no timeout — scout can run long
  }, (err) => {
    if (err) console.error(`[scout] exited with error for ${campaign}:`, err.message);
    else console.log(`[scout] completed for ${campaign}`);
  });
  console.log(`[scout] spawned PID ${child.pid} for campaign=${campaign}` + (keyword ? ` keyword="${keyword}"` : ''));
}

const sdk = new IMessageSDK({
  watcher: {
    pollInterval: 3000,
    excludeOwnMessages: true,
  },
});

const processed = new Set();

console.log('[tiktok-lookup] Starting iMessage watcher...');

await sdk.startWatching({
  onMessage: async (msg) => {
    if (processed.has(msg.id)) return;
    processed.add(msg.id);
    if (processed.size > 500) {
      const arr = [...processed];
      processed.clear();
      arr.slice(-250).forEach(id => processed.add(id));
    }

    // --- Scout command handler ---
    const scoutMatch = msg.text?.match(SCOUT_RE);
    if (scoutMatch) {
      const rawCampaign = scoutMatch[1];
      const keyword = scoutMatch[2]?.trim() || null;
      const sender = msg.sender;

      const campaign = resolveCampaign(rawCampaign);
      if (!campaign) {
        const available = listCampaigns().join(', ') || 'none';
        await sdk.send(sender, `Campaign '#${rawCampaign}' not found.\nAvailable: ${available}`);
        return;
      }

      if (keyword) {
        await sdk.send(sender, `Starting scout: #${campaign} "${keyword}"...`);
        spawnScout(campaign, keyword);
      } else {
        const pending = getPendingCount(campaign);
        if (pending === 0) {
          await sdk.send(sender, `No pending keywords for #${campaign}.`);
        } else {
          await sdk.send(sender, `Starting scout: #${campaign} (${pending} pending keyword${pending !== 1 ? 's' : ''})...`);
          spawnScout(campaign, null);
        }
      }
      return;
    }

    // --- TikTok URL handler ---
    const match = msg.text?.match(TIKTOK_URL_RE);
    if (!match) return;

    // Resolve short URL (tiktok.com/t/xxx) to handle via redirect + TikHub API
    let handle = match[1]; // direct @handle match
    if (!handle && match[2]) {
      try {
        const res = await fetch(`https://www.tiktok.com/t/${match[2]}`, {
          method: 'HEAD', redirect: 'follow',
        });
        // Try direct handle from redirect URL first
        const directHandle = res.url.match(/tiktok\.com\/@([\w.]+)\/video/i)?.[1];
        if (directHandle) {
          handle = directHandle;
        } else {
          // Redirect goes to /@/video/<id> without username — resolve via TikHub
          const videoId = res.url.match(/\/video\/(\d+)/i)?.[1];
          if (!videoId) { console.log(`[tiktok-lookup] Could not resolve short URL: ${match[0]}`); return; }
          const apiKey = process.env.TIKHUB_API_KEY;
          const detail = await fetch(
            `https://api.tikhub.io/api/v1/tiktok/web/fetch_post_detail?itemId=${videoId}`,
            { headers: { Authorization: `Bearer ${apiKey}` } }
          ).then(r => r.json());
          const h = detail?.data?.itemInfo?.itemStruct?.author?.uniqueId;
          if (!h) { console.log(`[tiktok-lookup] Could not resolve video author for ${videoId}`); return; }
          handle = h;
        }
      } catch (e) {
        console.error('[tiktok-lookup] Short URL resolve failed:', e.message);
        return;
      }
    }
    const sender = msg.sender;

    console.log(`[tiktok-lookup] TikTok URL detected from ${sender}: @${handle}`);

    // Acknowledge immediately
    try {
      await sdk.send(sender, `Looking up similar creators for @${handle}...`);
    } catch (e) {
      console.error('[tiktok-lookup] Failed to send ack:', e.message);
    }

    // Run lookup
    try {
      const result = execFileSync('uv', ['run', LOOKUP_PY, handle, sender, DATA_DIR], {
        timeout: 30000,
        env: { ...process.env },
      }).toString().trim();

      // Parse structured output: __HEADER__<text>\n__URLS__\n<url per line>
      const headerMatch = result.match(/^__HEADER__(.*?)(?:\n__URLS__\n([\s\S]*))?$/);
      if (headerMatch) {
        const header = headerMatch[1].trim();
        const urlBlock = headerMatch[2] ? '\n' + headerMatch[2].trim() : '';
        await sdk.send(sender, header + urlBlock);
      } else {
        await sdk.send(sender, result);
      }
      console.log(`[tiktok-lookup] Replied to ${sender} for @${handle}`);
    } catch (e) {
      const errMsg = e.stderr?.toString().trim() || e.message;
      console.error(`[tiktok-lookup] Lookup failed for @${handle}:`, errMsg);
      await sdk.send(sender, `Sorry, couldn't find similar creators for @${handle}. Try again later.`);
    }
  },

  onError: (err) => {
    console.error('[tiktok-lookup] Watcher error:', err.message);
  },
});

console.log('[tiktok-lookup] Watching for TikTok URLs in iMessages...');

process.on('SIGINT', async () => {
  sdk.stopWatching();
  await sdk.close();
  process.exit(0);
});

process.on('SIGTERM', async () => {
  sdk.stopWatching();
  await sdk.close();
  process.exit(0);
});

// Keep alive
await new Promise(() => {});

File: .claude/skills/tiktok-lookup/scripts/bot.py
# /// script
# dependencies = ["httpx", "python-dotenv"]
# ///
"""
bot.py — TikTok Similar Creator iMessage Bot (osascript edition)

Polls ~/Library/Messages/chat.db for new incoming messages.
Handles TikTok URL lookups and scout commands.
Sends replies via osascript — no Node.js required.

Run: uv run bot.py
"""

import os
import re
import sqlite3
import subprocess
import sys
import time
from pathlib import Path

# --- Path setup ---
SCRIPT_DIR = Path(__file__).resolve().parent
SKILL_DIR = SCRIPT_DIR.parent
ENV_PATH = SKILL_DIR.parents[1] / '.env'
DATA_DIR = SKILL_DIR.parents[2] / 'data'
LOOKUP_PY = SCRIPT_DIR / 'lookup.py'
SCOUT_PY = SKILL_DIR.parent / 'scout-api' / 'scripts' / 'scout.py'
KEYWORDS_HELPER_PY = SKILL_DIR.parent / 'scout-api' / 'scripts' / 'keywords_helper.py'
CAMPAIGNS_DIR = SKILL_DIR.parents[2] / 'context' / 'campaigns'
CHAT_DB = Path.home() / 'Library' / 'Messages' / 'chat.db'

# --- Load .env ---
if ENV_PATH.exists():
    for line in ENV_PATH.read_text().splitlines():
        line = line.strip()
        if line and not line.startswith('#') and '=' in line:
            k, _, v = line.partition('=')
            os.environ.setdefault(k.strip(), v.strip())

# --- Patterns ---
TIKTOK_URL_RE = re.compile(r'tiktok\.com/(?:@([\w.]+)|t/([\w]+))', re.IGNORECASE)
SCOUT_RE = re.compile(r'^scout\s+#(\w+)(?:\s+(.+))?$', re.IGNORECASE)

POLL_INTERVAL = 3  # seconds


# --- iMessage send via osascript ---
def send_imessage(recipient: str, text: str):
    # Escape backslashes and double quotes for AppleScript string
    safe = text.replace('\\', '\\\\').replace('"', '\\"')
    script = f'''tell application "Messages"
  set s to 1st service whose service type = iMessage
  send "{safe}" to buddy "{recipient}" of s
end tell'''
    try:
        subprocess.run(['osascript', '-e', script], timeout=15, capture_output=True)
    except Exception as e:
        print(f'[bot] send error to {recipient}: {e}', flush=True)


# --- Campaign helpers ---
def resolve_campaign(raw: str) -> str | None:
    try:
        dirs = [d for d in CAMPAIGNS_DIR.iterdir() if d.is_dir()]
        match = next((d.name for d in dirs if d.name.lower() == raw.lower()), None)
        return match
    except Exception:
        return None


def list_campaigns() -> list[str]:
    try:
        return [d.name for d in CAMPAIGNS_DIR.iterdir() if d.is_dir()]
    except Exception:
        return []


def get_pending_count(campaign: str) -> int:
    try:
        result = subprocess.run(
            ['uv', 'run', str(KEYWORDS_HELPER_PY), 'pending', campaign],
            capture_output=True, text=True, timeout=10,
        )
        return int(result.stdout.strip()) if result.returncode == 0 else 0
    except Exception:
        return 0


# --- Scout spawn ---
def spawn_scout(campaign: str, keyword: str | None, sender: str):
    args = ['uv', 'run', str(SCOUT_PY), campaign]
    if keyword:
        args.append(keyword)
    try:
        proc = subprocess.Popen(args, env=os.environ.copy(),
                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        print(f'[bot] spawned scout PID {proc.pid} campaign={campaign}' +
              (f' keyword="{keyword}"' if keyword else ''), flush=True)
    except Exception as e:
        print(f'[bot] scout spawn error: {e}', flush=True)
        send_imessage(sender, f'Failed to start scout for #{campaign}: {e}')


# --- Short URL resolution ---
def resolve_short_url(short_id: str) -> str | None:
    """Resolve tiktok.com/t/<id> to a handle via redirect + TikHub API if needed."""
    try:
        import httpx
        resp = httpx.head(f'https://www.tiktok.com/t/{short_id}',
                          follow_redirects=True, timeout=10)
        url = str(resp.url)
        # Direct handle in redirect URL
        m = re.search(r'tiktok\.com/@([\w.]+)/video', url, re.IGNORECASE)
        if m:
            return m.group(1)
        # No handle — resolve via TikHub post detail
        vid_m = re.search(r'/video/(\d+)', url)
        if not vid_m:
            return None
        video_id = vid_m.group(1)
        api_key = os.environ.get('TIKHUB_API_KEY')
        detail = httpx.get(
            f'https://api.tikhub.io/api/v1/tiktok/web/fetch_post_detail',
            params={'itemId': video_id},
            headers={'Authorization': f'Bearer {api_key}'},
            timeout=15,
        ).json()
        return detail.get('data', {}).get('itemInfo', {}).get('itemStruct', {}).get('author', {}).get('uniqueId')
    except Exception as e:
        print(f'[bot] short URL resolve error: {e}', flush=True)
        return None


# --- Message handler ---
def handle_message(sender: str, text: str):
    # Scout command
    scout_m = SCOUT_RE.match(text.strip())
    if scout_m:
        raw_campaign = scout_m.group(1)
        keyword = scout_m.group(2).strip() if scout_m.group(2) else None
        campaign = resolve_campaign(raw_campaign)
        if not campaign:
            available = ', '.join(list_campaigns()) or 'none'
            send_imessage(sender, f"Campaign '#{raw_campaign}' not found.\nAvailable: {available}")
            return
        if keyword:
            send_imessage(sender, f'Starting scout: #{campaign} "{keyword}"...')
            spawn_scout(campaign, keyword, sender)
        else:
            pending = get_pending_count(campaign)
            if pending == 0:
                send_imessage(sender, f'No pending keywords for #{campaign}.')
            else:
                send_imessage(sender, f'Starting scout: #{campaign} ({pending} pending keyword{"s" if pending != 1 else ""})...')
                spawn_scout(campaign, None, sender)
        return

    # TikTok URL
    url_m = TIKTOK_URL_RE.search(text)
    if not url_m:
        return

    handle = url_m.group(1)
    short_id = url_m.group(2)

    if not handle and short_id:
        handle = resolve_short_url(short_id)
        if not handle:
            print(f'[bot] could not resolve short URL t/{short_id}', flush=True)
            return

    print(f'[bot] TikTok URL from {sender}: @{handle}', flush=True)
    send_imessage(sender, f'Looking up similar creators for @{handle}...')

    try:
        result = subprocess.run(
            ['uv', 'run', str(LOOKUP_PY), handle, sender, str(DATA_DIR)],
            capture_output=True, text=True, timeout=30,
            env=os.environ.copy(),
        )
        output = result.stdout.strip()
        # Parse structured output: __HEADER__<text>\n__URLS__\n<urls>
        m = re.match(r'^__HEADER__(.*?)(?:\n__URLS__\n([\s\S]*))?$', output)
        if m:
            header = m.group(1).strip()
            urls = ('\n' + m.group(2).strip()) if m.group(2) else ''
            send_imessage(sender, header + urls)
        else:
            send_imessage(sender, output or 'No results.')
        print(f'[bot] replied to {sender} for @{handle}', flush=True)
    except Exception as e:
        print(f'[bot] lookup error for @{handle}: {e}', flush=True)
        send_imessage(sender, f"Sorry, couldn't find similar creators for @{handle}. Try again later.")


# --- DB polling ---
def poll(last_rowid: int) -> int:
    try:
        con = sqlite3.connect(f'file:{CHAT_DB}?mode=ro', uri=True,
                              check_same_thread=False, timeout=5)
        cur = con.execute('''
            SELECT m.rowid, m.text, h.id
            FROM message m
            JOIN handle h ON m.handle_id = h.rowid
            WHERE m.rowid > ?
              AND m.is_from_me = 0
              AND m.text IS NOT NULL
            ORDER BY m.rowid ASC
        ''', (last_rowid,))
        rows = cur.fetchall()
        con.close()
        for rowid, text, sender in rows:
            last_rowid = max(last_rowid, rowid)
            handle_message(sender, text)
    except Exception as e:
        print(f'[bot] db error: {e}', flush=True)
    return last_rowid


def get_latest_rowid() -> int:
    try:
        con = sqlite3.connect(f'file:{CHAT_DB}?mode=ro', uri=True, timeout=5)
        row = con.execute('SELECT MAX(rowid) FROM message').fetchone()
        con.close()
        return row[0] or 0
    except Exception:
        return 0


# --- Main ---
def main():
    print('[bot] Starting iMessage watcher (osascript edition)...', flush=True)
    last_rowid = get_latest_rowid()
    print(f'[bot] Watching from rowid {last_rowid}', flush=True)

    while True:
        last_rowid = poll(last_rowid)
        time.sleep(POLL_INTERVAL)


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('[bot] Stopped.', flush=True)
        sys.exit(0)
